File: [0322][Medium][Coin_Change]/Coin_Change_3.java
// leetcode time     cost : 15 ms
// leetcode memory   cost : 40.5 MB 
// Time  Complexity: O(S*n), S is the needed money amount
// Space Complexity: O(S)
// slolution 3, Dynamic programming - Bottom up,loop money amount with recursion
public class Solution {
  public int coinChange(int[] coins, int amount) {
    int max = amount + 1;
    int[] dp = new int[amount + 1];
    Arrays.fill(dp, max);
    dp[0] = 0;
    // loop the money amout in the outer iteration,then do the deeper recursion.
    for (int i = 1; i <= amount; i++) {
      for (int j = 0; j < coins.length; j++) {
        if (coins[j] <= i) {
          dp[i] = Math.min(dp[i], dp[i - coins[j]] + 1);
        }
      }
    }
    return dp[amount] > amount ? -1 : dp[amount];
  }
}

File: [0322][Medium][Coin_Change]/Coin_Change_3.py
# leetcode time     cost : 328 ms
# leetcode memory   cost : 13.6 MB 
# Time  Complexity: O(n)
# Space Complexity: O(n)
# slolution 2, Dynamic programming - Bottom up,loop money amount with Iterative DFS,
class Solution:
    def coinChange(self, coins, amount):
        coins.sort()
        stack = [(0, 0, len(coins))] # steps, accumulated
        min_steps = float('inf')
        while len(stack) != 0:
            steps, accumulated, sequence = stack.pop()
            if accumulated == amount:
                min_steps = min(min_steps, steps)
            if accumulated > amount or amount - accumulated > coins[sequence-1] * (min_steps-steps):
                continue
            for seq, coin in enumerate(coins[:sequence]):
                stack.append((steps+1, accumulated+coin, seq+1))
        return min_steps if min_steps != float('inf') else -1